File: swift/1694_ReformatPhoneNumber.swift
class Solution {
    func reformatNumber(_ number: String) -> String {
        var output = [String]()
        var num = number.replacingOccurrences(of: #"[\s-]"#, with: "", options: .regularExpression)
        
        var i = 0
        var idx = num.startIndex
        while i < num.count {
            var next = ""
            if i + 4 == num.count {
                for j in i..<i+2 {
                    next.append(num[idx])
                    idx = num.index(after: idx)
                }
                i += 2
                output.append(next)
            } else if i + 3 <= num.count {
                for j in i..<i+3 {
                    next.append(num[idx])
                    idx = num.index(after: idx)
                }
                i += 3
                output.append(next)
            } else {
                for j in i..<i+2 {
                    next.append(num[idx])
                    idx = num.index(after: idx)
                }
                i += 2
                output.append(next)
            }
        }
        
        return output.count == 1 ? output[0] : output.joined(separator: "-")
    }
}

File: cpp/1694_ReformatPhoneNumber.cpp
class Solution {
public:
    string reformatNumber(string number) {
        string num = "";
        for (char c : number) {
            if (c != ' ' && c != '-')
                num += c;
        }
        

        int l = num.length();
        string res = "";
        int idx = 0;
        while (l > 4 && idx < l) {
            res += num[idx];
            if (++idx % 3 == 0) {
                if (idx != l - 1) res += '-';
                if (l - idx <= 4) break;
            }
        }
        
        int r = l - idx;
        int j = 0;
        for (int i = idx; i < l; ++i) {
            if (r % 3 == 0 || r == 2) {
                res += num[i];
            } else {
                res += num[i];
                j++;
                if (j % 2 == 0 && i != l - 1)
                    res += '-';
            }
        }
        return res;
    }
};
